### PR TITLE
split: implementing separator option

### LINKED ISSUE #3192: split should support the -t/--separator option
```
‘-t separator’
‘--separator=separator’

    Use character separator as the record separator instead of the default newline character (ASCII LF). To specify ASCII NUL as the separator, use the two-character string ‘\0’, e.g., ‘split -t '\0'’.
```

Additionally, it would be useful to support multiple character separators, e.g. '\r\n'

### PR DESCRIPTION
**Pull request by zhitkoff**
Implementing `-t` / `--separator` option + some `--help` improvements

Fixes #3192

### CODE REVIEW

**Comment by sylvestre** [comment]
> Congrats! The gnu test tests/split/record-sep is no longer failing!

Kudos!

File: src/uu/split/src/split.rs [hunk 1]
```diff
@@ -65,6 +67,86 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+/// However it will leave other values untouched, so
+/// `split -t: -t: -t,` will become `split -t: -t,` and Clap will panic next as it should
+fn handle_multiple_separator_options(args: &[OsString]) -> Vec<OsString> {
+    let mut separator_values: Vec<&str> = vec![];
+    let mut preceding_sep_opt_req_value = false;
+    let mut i = -1;
+    let mut slices_to_remove: Vec<i32> = vec![];
+
+    // collect slices to remove
+    for os_slice in args {
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:85]: you probably can do the i this way:

    for (i, os_slice) in args.iter().enumerate() {

File: src/uu/split/src/split.rs [hunk 2]
```diff
@@ -65,6 +67,86 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
+        i += 1;
+        if let Some(slice) = os_slice.to_str() {
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:87]: this feels quite complex, maybe you could do something like:

         let (prefix, offset) = if slice.starts_with("-t") {
                ("-t", 2)
            } else if slice.starts_with("--separator=") {
                ("--separator=", 12)
            } else if slice.starts_with("--separator") {
                ("--separator", 11)
            } else {
                ("", 0)
            };

File: src/uu/split/src/split.rs [hunk 3]
```diff
@@ -783,6 +888,26 @@ impl Settings {
...
+                        Some(b'\n')
+                    } else if s == "\\0" {
+                        Some(b'\0')
+                    } else {
+                        return Err(SettingsError::MultiCharacterSeparator(s.to_owned()));
+                    }
+                }
+            }
+            _ => None,
+        };
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:909]: can probably be simplified by something like:
```suggestion
let separator = matches.get_one::<String>(OPT_SEPARATOR).and_then(|s| {
    match s.as_str() {
        "\\n" => Some(b'\n'),
        "\\0" => Some(b'\0'),
        _ if s.chars().count() == 1 => Some(s.as_bytes()[0]),
        _ => return Err(SettingsError::MultiCharacterSeparator(s.to_owned())),
    }
});

```

File: src/uu/split/src/split.rs (hunk 1 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:85]: possibly, yes. I actually switched to the current form based on clippy suggestion. Will see if this can be refactored further

File: src/uu/split/src/split.rs (hunk 3 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:909]: fantastic, thank you! will refactor

File: src/uu/split/src/split.rs (hunk 2 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:87]: this does not take into account concatenated value vs value separated by whitespace, but I think it can be added into this format for each if/if else. Will try to refactor.
thanks!

Comment (zhitkoff) [reply src/uu/split/src/split.rs:87]: unfortunately it does not look like it can be much simplified here, since both cases need to be covered - concatenated values like `--separator=:` or `-t,` as well as values separated by white space like `--separator :` or `-t ,` 
We cannot just get (prefix, offset) - it would only work for concatenated case.

Comment (zhitkoff) [reply src/uu/split/src/split.rs:87]: Btw - this function exists only to pass GNU test for split separator

**Comment by sylvestre** [comment]
don't bother :) I can do it in Github directly

File: src/uu/split/src/split.rs [hunk 4]
```diff
@@ -65,6 +67,82 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
 
+/// If the same separator (with the same value) was used multiple times - `split` should NOT fail
+/// If the separator was used multiple times but with different values - `split` should fail
+/// Unfortunately Clap cannot support both conditions simultaneously, so it is setup to fail on the second condition
+/// For the first condition - this function will find if there are multiple separator options with the same value
+/// and remove duplicates, leaving only one
+/// i.e. `split -t: -t:` will become `split -t:`
+/// However it will leave other values untouched, so
+/// `split -t: -t: -t,` will become `split -t: -t,` and Clap will panic next as it should
+fn handle_multiple_separator_options(args: &[OsString]) -> Vec<OsString> {
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:78]: do you think you could write pure rust unit tests just for this function?

Comment (zhitkoff) [reply src/uu/split/src/split.rs:78]: should be doable, yes. Before I do though - isn't it already 100% covered by tests from `tests/by_util/test_split.rs` ?

File: src/uu/split/src/split.rs [hunk 5]
```diff
@@ -65,6 +67,82 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
...
 }
 
+/// If the same separator (with the same value) was used multiple times - `split` should NOT fail
+/// If the separator was used multiple times but with different values - `split` should fail
+/// Unfortunately Clap cannot support both conditions simultaneously, so it is setup to fail on the second condition
+/// For the first condition - this function will find if there are multiple separator options with the same value
+/// and remove duplicates, leaving only one
+/// i.e. `split -t: -t:` will become `split -t:`
+/// However it will leave other values untouched, so
+/// `split -t: -t: -t,` will become `split -t: -t,` and Clap will panic next as it should
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:77]: Can't we still use clap here with some custom post-processing?

I'm thinking that it should be fairly simple to collect multiple values for the separator and then check that all given values are the same.

Would that give a different result?

The problem with an approach like this one is that you theoretically need to keep _all_ options into account. For example what if the `-t` is the argument to the `--filter` flag. It's unlikely, but possible and would be difficult to fix. Two separate pre-processing steps is also just very messy to deal with. Whenever we get a bad output from `clap` now, it could be clap, `handle_obsolete` or this function that messed something up.

File: src/uu/split/src/split.rs [hunk 6]
```diff
@@ -208,13 +287,18 @@ fn handle_preceding_options(
...
     // following slice should be treaded as value for this option
     // even if it starts with '-' (which would be treated as hyphen prefixed value)
-    *preceding_short_opt_req_value =
-        slice == "-b" || slice == "-C" || slice == "-l" || slice == "-n" || slice == "-a";
+    *preceding_short_opt_req_value = slice == "-b"
+        || slice == "-C"
+        || slice == "-l"
+        || slice == "-n"
+        || slice == "-a"
+        || slice == "-t";
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:301]:
```suggestion
    *preceding_short_opt_req_vlaue = matches!(slice, "-b" | "-C" | "-l" | "-n" | "-a" | "-t");
```

File: src/uu/split/src/split.rs [hunk 7]
```diff
@@ -340,6 +425,7 @@ pub fn uu_app() -> Command {
                     OPT_HEX_SUFFIXES,
                     OPT_HEX_SUFFIXES_SHORT
                 ])
+                .value_name("FROM")
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:428]: Nice little additions here!

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
I think this is a great attempt, but I would like to see if we can still use clap somehow, because this becomes a bit messy. If that's not possible, then this is alright.

File: src/uu/split/src/split.rs (hunk 4 above)
Comment (sylvestre) [reply src/uu/split/src/split.rs:78]: They are but harder to debug in case of regression :)

File: src/uu/split/src/split.rs [hunk 8]
```diff
@@ -783,6 +884,26 @@ impl Settings {
                 }
             }
         }
+
+        // Make sure that separator is only one UTF8 character (if specified)
+        let mut multi_char_sep = "";
+        let separator = matches
+            .get_one::<String>(OPT_SEPARATOR)
+            .and_then(|s| match s.as_str() {
+                "\\n" => Some(b'\n'),
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/split.rs:893]: yep, it looks like we only need to support '\0'. thanks!

File: src/uu/split/split.md [hunk 9]
```diff
@@ -11,3 +11,15 @@ Create output files containing consecutive or interleaved sections of input
...
+Units are K,M,G,T,P,E,Z,Y,R,Q (powers of 1024) or KB,MB,... (powers of 1000).
+Binary prefixes can be used, too: KiB=K, MiB=M, and so on.
+
+CHUNKS may be:
+  N       split into N files based on size of input
+  K/N     output Kth of N to stdout
+  l/N     split into N files without splitting lines/records
+  l/K/N   output Kth of N to stdout without splitting lines/records
+  r/N     like 'l' but use round robin distribution
+  r/K/N   likewise but only output Kth of N to stdout
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/split.md:25]: yep, it would look better online with list, but I am afraid it will not look good in the terminal, i.e. it will look like this:

```
CHUNKS may be:

- N       split into N files based on size of input
- K/N     output Kth of N to stdout
- l/N     split into N files without splitting lines/records
- l/K/N   output Kth of N to stdout without splitting lines/records
- r/N     like 'l' but use round robin distribution
- r/K/N   likewise but only output Kth of N to stdout
```

or 

```
CHUNKS may be:

* N       split into N files based on size of input
* K/N     output Kth of N to stdout
* l/N     split into N files without splitting lines/records
* l/K/N   output Kth of N to stdout without splitting lines/records
* r/N     like 'l' but use round robin distribution
* r/K/N   likewise but only output Kth of N to stdout
```

which is not the same as in GNU version and could also be misleading/confusing - implying that CHUNKS should be entered with leading `-` or `*` followed by space and value or some such

File: src/uu/split/src/split.rs (hunk 5 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:77]: yes, we could use Clap's `overrides_with()` pointed to itself - in that case no matter how many times `-t` and/or `--separator` is used - the last one would win. However, that will break GNU test as it checks for both cases mentioned in the comment section for this function, i.e. this specifically will NOT cause Clap to fail with `overrides_with()` pointed to itself, but GNU version expects it to fail:
```
split -ta -tb </dev/null
```

File: src/uu/split/split.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/split/split.md:25]: Hmm yeah, I understand your hesitation. I think we've done this before in some other utils too and it's generally fine to use the markdown syntax. Maybe some day we'll add some additional pre-processing for the help text :)

File: src/uu/split/src/split.rs (hunk 6 above)
Comment (tertsdiepraam) [reply src/uu/split/src/split.rs:301]: Feel free to leave this btw, since you just extended what was already there.

File: src/uu/split/src/split.rs (hunk 5 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:77]: > The problem with an approach like this one is that you theoretically need to keep _all_ options into account. For example what if the `-t` is the argument to the `--filter` flag. It's unlikely, but possible and would be difficult to fix. Two separate pre-processing steps is also just very messy to deal with. Whenever we get a bad output from `clap` now, it could be clap, `handle_obsolete` or this function that messed something up.

In this example - it will not break as this function will only remove arg slices for complete duplicates of separator/value pairs (concatenated or not), i.e. in case of `split --filter -t in` nothing will be altered.

Comment (tertsdiepraam) [reply src/uu/split/src/split.rs:77]: Ah sorry, I didn't mean `overrides`, but just collecting multiple values, for example with [`ArgAction::append`](https://docs.rs/clap/latest/clap/enum.ArgAction.html#variant.Append). You'd get a `Vec` of all the values that you can check.

File: src/uu/split/src/split.rs [hunk 10]
```diff
@@ -1019,7 +1141,11 @@ impl<'a> Write for LineChunkWriter<'a> {
         // corresponds to the current chunk number.
         let mut prev = 0;
         let mut total_bytes_written = 0;
-        for i in memchr::memchr_iter(b'\n', buf) {
+        let sep = match &self.settings.separator {
+            Some(s) => *s,
+            _ => b'\n',
+        };
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/split.rs:1147]: thank you! yes, it would be better - will do.

File: src/uu/split/split.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/split.md:25]: sounds good. What would be the preference for the markdown list character then (to stay consistent with other utils)? Should it be `-` or `+` or `*` ?

File: src/uu/split/src/split.rs [hunk 11]
```diff
@@ -783,6 +884,26 @@ impl Settings {
...
         }
+
+        // Make sure that separator is only one UTF8 character (if specified)
+        let mut multi_char_sep = "";
+        let separator = matches
+            .get_one::<String>(OPT_SEPARATOR)
+            .and_then(|s| match s.as_str() {
+                "\\n" => Some(b'\n'),
+                "\\0" => Some(b'\0'),
+                _ if s.chars().count() == 1 => Some(s.as_bytes()[0]),
```
[the remark this answers is not in the document]
Comment (zhitkoff) [reply src/uu/split/src/split.rs:895]: yep, will refactor to all bytes/u8 as separator must be u8 for i/o functions later on

File: src/uu/split/split.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/split/split.md:25]: Let's see, we used `-` before (e.g. https://github.com/uutils/coreutils/pull/5260/files) so let's do that. Thanks!

File: src/uu/split/src/split.rs (hunk 5 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:77]: fantastic suggestion, thank you. Looks like this will work and with less code too :)

File: src/uu/split/src/split.rs (hunk 4 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:78]: turns out this function is not necessary - refactored using Clap `ArgAction::append` - kudos to @tertsdiepraam for suggestion :)

**Comment by zhitkoff** [comment]
@sylvestre @tertsdiepraam looking for your input - would the latest changes work or does it need more refactoring?

File: src/uu/split/src/split.rs [hunk 12]
```diff
@@ -743,6 +764,9 @@ impl fmt::Display for SettingsError {
             Self::Strategy(e) => e.fmt(f),
             Self::SuffixNotParsable(s) => write!(f, "invalid suffix length: {}", s.quote()),
             Self::SuffixTooSmall(i) => write!(f, "the suffix length needs to be at least {i}"),
+            Self::MultiCharacterSeparator(s) => {
+                write!(f, "multi-character separator '{}'", s.quote())
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:768]: Wouldn't this quote the strings twice?

File: src/uu/split/src/split.rs [hunk 13]
```diff
@@ -783,6 +807,29 @@ impl Settings {
...
+        // If the same separator (the same value) was used multiple times - `split` should NOT fail
+        // If the separator was used multiple times but with different values (not all values are the same) - `split` should fail
+        let separator = match matches.get_many::<String>(OPT_SEPARATOR) {
+            Some(sep_values) => match sep_values.map(|s| s.to_string()).reduce(|cur, nxt| {
+                if cur == nxt {
+                    nxt
+                } else {
+                    cur + &nxt
+                }
+            }) {
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:822]: This is a bit hard to read at the moment. Can we maybe do the reduce on a separate line?

File: src/uu/split/src/split.rs [hunk 14]
```diff
@@ -783,6 +807,29 @@ impl Settings {
...
+                Some(s) => match s.as_str() {
+                    "\\0" => b'\0',
+                    _ if s.as_bytes().len() == 1 => s.as_bytes()[0],
+                    _ => return Err(SettingsError::MultiCharacterSeparator(s.to_owned())),
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:826]: There's a subtlety here in the GNU implementation:
```
> split -t1 -t2
split: multiple separator characters specified
> split -t12
split: multi-character separator ‘12’
```
I don't really mind this version though. I just wanted to note it's not quite the same.

File: src/uu/split/src/split.rs [hunk 15]
```diff
@@ -783,6 +807,29 @@ impl Settings {
                 }
             }
         }
+
+        // Make sure that separator is only one UTF8 character (if specified)
+        // defaults to '\n' - newline character
+        // If the same separator (the same value) was used multiple times - `split` should NOT fail
+        // If the separator was used multiple times but with different values (not all values are the same) - `split` should fail
+        let separator = match matches.get_many::<String>(OPT_SEPARATOR) {
+            Some(sep_values) => match sep_values.map(|s| s.to_string()).reduce(|cur, nxt| {
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:816]: A `fold` might be more appropriate here, to avoid the `Option` in the return type.

You could also write the check like this instead of the reduce:
```rust
let first = sep_values.next();
match first {
    Some(first) => {
        sep_values.all(|s| s == first);
        // match on the value here
    }
    None => b'\n',
}
```
I'll leave it up to you what you think looks best in the end.

**Review by tertsdiepraam — COMMENTED** [review]
Looking great! Just a few small suggestions.

File: src/uu/split/src/split.rs (hunk 12 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:768]: yep, good catch, thanks

File: src/uu/split/src/split.rs (hunk 15 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs:816]: just pushed a new version - still using reduce, but since we can assume that Clap's get_many() call should return non-empty vector inside Some(ValuesRef<'_,String>) - it is safe to just unwrap and skip several match blocks one within another - just have one.

File: src/uu/split/src/split.rs [hunk 16]
```diff
@@ -783,6 +807,25 @@ impl Settings {
...
+
+        // Make sure that separator is only one UTF8 character (if specified)
+        // defaults to '\n' - newline character
+        // If the same separator (the same value) was used multiple times - `split` should NOT fail
+        // If the separator was used multiple times but with different values (not all values are the same) - `split` should fail
+        let separator = match matches.get_many::<String>(OPT_SEPARATOR) {
+            Some(sep_values) => match sep_values
+                .map(|s| s.to_string())
+                .reduce(|cur, nxt| if cur == nxt { nxt } else { cur + &nxt })
+                .unwrap() // it is safe to just unwrap here since Clap should not return empty ValuesRef<'_,String> in the option from get_many() call
```
Comment (tertsdiepraam) [inline src/uu/split/src/split.rs:819]: If you're gonna unwrap, this is all you need I think.
```rust
let first = sep_values.next().unwrap():
if !sep_values.all(|s| s == first) {
    // multiple values given error
}
match first.as_str() {
   ...
}
```
I find the `reduce` a bit hard to follow on second glance, because it is unclear what `cur` represents. It's first a single value and then multiple values glued together, which makes it a bit confusing.

File: src/uu/split/src/split.rs (hunk 14 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs]: in latest refactor - added new error variant + tests to distinguish the two

File: src/uu/split/src/split.rs (hunk 16 above)
Comment (zhitkoff) [reply src/uu/split/src/split.rs]: no problem, pushed updated version

**Review by tertsdiepraam — APPROVED** [review]
Very nice! Thanks!
